Google.Api.Generator/Generation/MetadataGenerator.cs

```diff
@@ -0,0 +1,70 @@
+﻿using Google.Gapic.Metadata;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+
+namespace Google.Api.Generator.Generation
+{
+    internal static class MetadataGenerator
+    {
+        private const string GapicMetadataSchemaVersion = "1.0";
+        private const string GapicMetadataFileDescription = "This file maps proto services/RPCs to the corresponding library clients/methods";
+        private const string GapicMetadataLanguage = "csharp";
+
+        private const string TransportKeyGrpc = "grpc";
+
+        internal static string GenerateGapicMetadataJson(List<ServiceDetails> allServiceDetails)
+        {
+            
+            var gapicMetadata = new GapicMetadata
+            {
+                Schema = GapicMetadataSchemaVersion,
+                Comment = GapicMetadataFileDescription,
+                Language = GapicMetadataLanguage,
+                ProtoPackage = allServiceDetails.First().Package,
+                LibraryPackage = allServiceDetails.First().Namespace
+            };
+
+            foreach (var serviceDetails in allServiceDetails)
+            {
+                gapicMetadata.Services[serviceDetails.ServiceName] = ServiceForTransportMetadata(serviceDetails);
+            }
+
+            var gapicMetadataJson = JsonSerializer.Serialize(
+                gapicMetadata,
+                new JsonSerializerOptions
+                {
+                    WriteIndented = true,
+                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+                    DictionaryKeyPolicy = JsonNamingPolicy.CamelCase
+                });
+
+            return gapicMetadataJson + "\n"; //trailing file newline to please github
+        }
+
+        private static GapicMetadata.Types.ServiceForTransport ServiceForTransportMetadata(ServiceDetails serviceDetails)
+        {
+            var serviceForTransportMetadata = new GapicMetadata.Types.ServiceForTransport();
+            
+            serviceForTransportMetadata.Clients[TransportKeyGrpc] = ServiceAsClientMetadata(serviceDetails);
+            return serviceForTransportMetadata;
+        }
+
+        private static GapicMetadata.Types.ServiceAsClient ServiceAsClientMetadata(ServiceDetails serviceDetails)
+        {
+            var serviceAsClient = new GapicMetadata.Types.ServiceAsClient
+            {
+                LibraryClient = serviceDetails.ClientImplTyp.FullName
+            };
+
+            foreach (var methodDetails in serviceDetails.Methods)
+            {
+                var methodsList = new GapicMetadata.Types.MethodList();
+                methodsList.Methods.AddRange(new[] { methodDetails.SyncMethodName, methodDetails.AsyncMethodName });
+                serviceAsClient.Rpcs[methodDetails.MethodName] = methodsList;

```

Comment:
I think you should be able to do something like:
```suggestion
                serviceAsClient.Rpcs[methodDetails.MethodName] = new GapicMetadata.Types.MethodList
                {
                    methodDetails.SyncMethodName,
                    methodDetails.AsyncMethodName
                };

```

Reply:
That's about what I initially tried to do but `Methods` has no setter, and neither are other properties like `Rpcs` or `Clients`

Reply:
They're repeated fields or maps - no setter, but the syntax above with the `{ ... }` calls Add, which is overloaded to include a *range* of values.

Reply:
now that I know about the `={}` on proto collection classes, I was able to make this whole class much prettier (it will be even prettier when target-typed object creation hits.

Reply:
👍